### PR TITLE
Multiple improvements by CodeRush Static Analysis

### PR DESCRIPTION
**Pull request by Himura2la**
Hi guys!
I have checked the PowerShell solution with the [CodeRush](https://www.devexpress.com/products/coderush/) Static Analyzer and fixed several code issues. Here's the summary:

- The most common issue is [CRR0012: Logical OR expression has opposite operand](https://documentation.devexpress.com/CodeRushForRoslyn/118112/Static-Code-Analysis/Analyzers-Library/CRR0012-Logical-OR-expression-has-opposite-operands), such expressions can be simplified.
- In [Words.cs](https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.PSReadLine/Words.cs#L112) our analyzers detect [CRR0025: This conditional's code block does not appear to be reachable](https://documentation.devexpress.com/CodeRushForRoslyn/118235/Static-Code-Analysis/Analyzers-Library/CRR0025-Unreachable-conditional-code-block-the-inversed-condition-is-already-satisfied), it's obviously false positive, but these searching `while` loops caught my attention. They look very similar to nested conditionals and I decided to fold them into for loops without bodies, they look clearer in this form.
- Several places had [CRR0008: If-block matches else-block](https://documentation.devexpress.com/CodeRushForRoslyn/118106/Static-Code-Analysis/Analyzers-Library/CRR0008-If-block-matches-else-block), I saved the original conditionals in comments in case they need in future
- I have found the **CRR0029: The ConfigureAwait(true) is called implicitly** issue. The simple delay does not require saving the synchronization context, plus it's better always add an explicit `ConfigureAwait()` call to `await`ed calls.
- In [WSMan](https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.WSMan.Management/ConfigProvider.cs#L2584) I found [CRR0017: Expression checks 'nodelist' for null after usage](https://documentation.devexpress.com/CodeRushForRoslyn/118132/Static-Code-Analysis/Analyzers-Library/CRR0017-Null-check-follows-usage) and one more small issue, maybe I should add another pull request for that repository.

I hope this will help, here's the full report: [CodeIssuesReport.txt](https://github.com/PowerShell/PowerShell/files/1387705/CodeIssuesReport.txt). Thanks for being open source!

### CODE REVIEW

File: src/Microsoft.PowerShell.PSReadLine/Words.cs [hunk 1]
```diff
@@ -143,24 +131,10 @@ private int FindNextWordPoint(string wordDelimiters)
...
 
-            while (i < _singleton._buffer.Length)
-            {
-                if (InWord(i, wordDelimiters))
-                {
-                    break;
-                }
-                i += 1;
-            }
+            for (; i < _singleton._buffer.Length && !InWord(i, wordDelimiters); i++) ;
```
Comment (PaulHigin) [inline src/Microsoft.PowerShell.PSReadLine/Words.cs]: I feel this is easier to read (with the additional parentheses):
for (; (i < _singleton._buffer.Length) && !InWord(I, wordDelimiters); I++);

File: src/System.Management.Automation/engine/DataStoreAdapter.cs [hunk 2]
```diff
@@ -856,16 +856,9 @@ public bool Equals(PSDriveInfo drive)
...
-                {
-                    // Since both drives are null, they are equal
-                    return false;
-                }
-                else
-                {
-                    // Since drive1 is null it is less than drive2 which is not null
-                    return false;
-                }
+                // Since both drives are null, they are equal
```
Comment (PaulHigin) [inline src/System.Management.Automation/engine/DataStoreAdapter.cs]: This doesn't look right.  I believe it should be.
return (drive2Object != null);

File: src/System.Management.Automation/engine/DataStoreAdapterProvider.cs [hunk 3]
```diff
@@ -405,11 +405,7 @@ protected ProviderInfo(ProviderInfo providerInfo)
                 throw PSTraceSource.NewArgumentException("name");
             }
 
-
-            if (String.IsNullOrEmpty(name))
-            {
-                throw PSTraceSource.NewArgumentException("name");
-            }
+            // FIXME: Other checks?
```
Comment (PaulHigin) [inline src/System.Management.Automation/engine/DataStoreAdapterProvider.cs]: Please remove this comment and verify that helpfile, description, psSnapIn parameters can be null.

File: src/System.Management.Automation/engine/Modules/AnalysisCache.cs [hunk 4]
```diff
@@ -634,14 +634,14 @@ public void QueueSerialization()
                     // Wait a while before assuming we've finished the updates,
                     // writing the cache out in a timely matter isn't too important
                     // now anyway.
-                    await Task.Delay(10000);
+                    await Task.Delay(10000).ConfigureAwait(false);
```
Comment (PaulHigin) [inline src/System.Management.Automation/engine/Modules/AnalysisCache.cs]: It is not clear to me that this is needed.  Isn't the Task always configured as "false" be default?

File: src/System.Management.Automation/security/Authenticode.cs [hunk 5]
```diff
@@ -619,9 +619,9 @@ private static X509Certificate2 GetCertFromChain(IntPtr pSigner)
                 }
             }
 
-            Diagnostics.Assert(((error == 0) && (signature != null)) || (error != 0), "GetSignatureFromWintrustData: general crypto failure");
+            Diagnostics.Assert(error != 0 || signature != null, "GetSignatureFromWintrustData: general crypto failure");
 
-            if ((signature == null) && (error != 0))
+            if (signature == null && error != 0)
```
Comment (PaulHigin) [inline src/System.Management.Automation/security/Authenticode.cs]: I feel the code is more readable with parentheses back in.  I am not sure what our coding guidelines say.

**Review by PaulHigin — CHANGES_REQUESTED** [review]
Thanks for finding and fixing these!

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostRawUserInterface.cs [hunk 6]
```diff
@@ -741,8 +741,7 @@ public override
                             if (((ConsoleControl.InputRecordEventTypes)inputRecords[0].EventType) ==
                                 ConsoleControl.InputRecordEventTypes.KEY_EVENT)
                             {
-                                Dbg.Assert((inputRecords[0].KeyEvent.KeyDown && inputRecords[0].KeyEvent.RepeatCount != 0) ||
-                                    !inputRecords[0].KeyEvent.KeyDown,
+                                Dbg.Assert(inputRecords[0].KeyEvent.RepeatCount != 0 || !inputRecords[0].KeyEvent.KeyDown,
```
Comment (lzybkr) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostRawUserInterface.cs]: This changes the order of evaluation - probably safe here, but before, `RepeatCount` would not be evaluated if `KeyDown` was false.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostRawUserInterface.cs [hunk 7]
```diff
@@ -867,8 +866,7 @@ public override
                     if (((ConsoleControl.InputRecordEventTypes)inputRecords[i].EventType) ==
                             ConsoleControl.InputRecordEventTypes.KEY_EVENT)
                     {
-                        Dbg.Assert((inputRecords[i].KeyEvent.KeyDown && inputRecords[i].KeyEvent.RepeatCount != 0) ||
-                            !inputRecords[i].KeyEvent.KeyDown,
+                        Dbg.Assert(inputRecords[i].KeyEvent.RepeatCount != 0 || !inputRecords[i].KeyEvent.KeyDown,
```
Comment (lzybkr) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostRawUserInterface.cs]: This changes the order of evaluation - probably safe here, but before, RepeatCount would not be evaluated if KeyDown was false.

File: src/Microsoft.PowerShell.PSReadLine/Words.cs (hunk 1 above)
Comment (lzybkr) [reply src/Microsoft.PowerShell.PSReadLine/Words.cs]: Submit a PR to https://github.com/lzybkr/PSReadLine for any changes in the PSReadLine directory.

Comment (lzybkr) [reply src/Microsoft.PowerShell.PSReadLine/Words.cs]: And I'll likely reject the change - In general I'm not a fan of empty loop bodies.

File: src/Microsoft.WSMan.Management/WsManHelper.cs [hunk 8]
```diff
@@ -496,19 +496,13 @@ internal string ProcessInput(IWSManEx wsman, string filepath, string operation,
...
-                tmpNs = StripParams(resUri);
-            else
-            {
-                //tmpNs = StripParams(resUri) + ".xsd";
-                //This was reported by Intel as an interop issue. So now we are not appending a .xsd in the end.
-                tmpNs = StripParams(resUri);
-            }
+            //if (!(resUri.ToLowerInvariant().Contains(URI_IPMI) || resUri.ToLowerInvariant().Contains(URI_WMI)))
+            //    tmpNs += ".xsd";
+            //This was reported by Intel as an interop issue. So now we are not appending a .xsd in the end.
```
Comment (lzybkr) [inline src/Microsoft.WSMan.Management/WsManHelper.cs]: This comment can probably be deleted.

File: src/System.Management.Automation/engine/Modules/AnalysisCache.cs (hunk 4 above)
Comment (lzybkr) [reply src/System.Management.Automation/engine/Modules/AnalysisCache.cs]: It seems right though, so I think it's good to keep it.

File: src/System.Management.Automation/engine/remoting/client/remotepipeline.cs [hunk 9]
```diff
@@ -912,8 +912,7 @@ private void InitPowerShellForConnect(bool syncCall)
 
             // Check to see if this pipeline already exists in the runspace.
             RemotePipeline currentPipeline = (RemotePipeline)((RemoteRunspace)_runspace).GetCurrentlyRunningPipeline();
-            if (currentPipeline == null ||
-                currentPipeline != null && !ReferenceEquals(currentPipeline, this))
+            if (currentPipeline == null || !ReferenceEquals(currentPipeline, this))
```
Comment (lzybkr) [inline src/System.Management.Automation/engine/remoting/client/remotepipeline.cs]: I think `currentPipeline == null` is redundant with `ReferenceEquals` here.

File: src/System.Management.Automation/security/Authenticode.cs (hunk 5 above)
Comment (lzybkr) [reply src/System.Management.Automation/security/Authenticode.cs]: In this area, coding guidelines say - be consistent, but we also ask for no formatting changes mixed with other changes.

---
In reply to: [144931951](https://github.com/PowerShell/PowerShell/pull/5132#discussion_r144931951) [](ancestors = 144931951)

**Comment by TravisEz13** [comment]
@Himura2la What is your status on signing the CLA?

File: src/Microsoft.PowerShell.CoreCLR.Eventing/DotNetCode/Eventing/Reader/EventLogReader.cs [hunk 10]
```diff
@@ -305,16 +305,9 @@ public void Seek(SeekOrigin origin, long offset)
                     }
                     else
                     {
-                        //if inside the current buffer, we still cannot read the events, as the handles.
+                        //if inside the current buffer (_currentIndex + offset >= 0), we still cannot read the events, as the handles.
```
[the remark this answers is not in the document]
Comment (Himura2la) [reply src/Microsoft.PowerShell.CoreCLR.Eventing/DotNetCode/Eventing/Reader/EventLogReader.cs]: OK, removing the whole comment

File: src/Microsoft.PowerShell.PSReadLine/Words.cs (hunk 1 above)
Comment (Himura2la) [reply src/Microsoft.PowerShell.PSReadLine/Words.cs]: I reverted the file, this change is a anyway just a preference. Maybe I'll try to push it to the right repository. thanks for the link.

File: src/System.Management.Automation/engine/DataStoreAdapter.cs (hunk 2 above)
Comment (Himura2la) [reply src/System.Management.Automation/engine/DataStoreAdapter.cs]: Oh, thanks, I didn't notice that the refactoring can go even further

File: src/System.Management.Automation/engine/Modules/AnalysisCache.cs (hunk 4 above)
Comment (Himura2la) [reply src/System.Management.Automation/engine/Modules/AnalysisCache.cs]: @PaulHigin As far as I know, the default option is `'true'`.

File: src/System.Management.Automation/engine/DataStoreAdapterProvider.cs (hunk 3 above)
Comment (Himura2la) [reply src/System.Management.Automation/engine/DataStoreAdapterProvider.cs]: Sorry, I think I didn't catch whether I should ensure the `helpfile`, `description` and `psSnapIn` is not null or just remove the comment.

File: src/System.Management.Automation/security/Authenticode.cs (hunk 5 above)
Comment (Himura2la) [reply src/System.Management.Automation/security/Authenticode.cs]: I reverted this. I should have keep myself from removing brackets everywhere, sorry.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostRawUserInterface.cs (hunk 6 above)
Comment (Himura2la) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostRawUserInterface.cs]: I swapped them, thanks for noticing. It seems really safe assuming there's no null-checks

File: src/Microsoft.WSMan.Management/WsManHelper.cs (hunk 8 above)
Comment (Himura2la) [reply src/Microsoft.WSMan.Management/WsManHelper.cs]: OK, I also inlined the `tmpNs` variable. I hope it's OK

File: src/System.Management.Automation/engine/remoting/client/remotepipeline.cs (hunk 9 above)
Comment (Himura2la) [reply src/System.Management.Automation/engine/remoting/client/remotepipeline.cs]: Seems like.

**Comment by Himura2la** [comment]
I'm done with the changes and CLA. The only thing left is extra checks [here](https://github.com/PowerShell/PowerShell/pull/5132/files/b829325d8d3016242942f58c5da4293698fdc549#diff-994563ab8f2b9856ba8449395e02847e)

**Review by lzybkr — APPROVED** [review]

**Comment by TravisEz13** [comment]
@PaulHigin Just a reminder to update your review.

Restarted macOS due to random brew download failure.

**Review by PaulHigin — APPROVED** [review]
LGTM

File: src/System.Management.Automation/engine/DataStoreAdapterProvider.cs (hunk 3 above)
Comment (TravisEz13) [reply src/System.Management.Automation/engine/DataStoreAdapterProvider.cs]: @Himura2la Can you either remove this comment or file an issue and update the comment to reference the issue?

**Review by TravisEz13 — CHANGES_REQUESTED** [review]

**Review by TravisEz13 — APPROVED** [review]

**Comment by TravisEz13** [comment]
I'm following up on one issue.  Hopefully there will be no more delays once I have the answer.

**Comment by Himura2la** [comment]
@TravisEz13, I've removed the comment on in my second commit

**Comment by TravisEz13** [comment]
@Himura2la Yeah, that is not the issue.

**Comment by iSazonov** [comment]
Is the PR ready to merge?

**Comment by Himura2la** [comment]
I wonder

**Comment by SteveL-MSFT** [comment]
@TravisEz13 is this ready to merge?

**Pull request merged**